pandas/_libs/groupby.pyx

```diff
@@ -117,6 +121,98 @@ def group_last_object(ndarray[object, ndim=2] out,
             else:
                 out[i, j] = resx[i, j]
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def group_rank_object(ndarray[float64_t, ndim=2] out,
+                      ndarray[object, ndim=2] values,
+                      ndarray[int64_t] labels,
+                      bint is_datetimelike, **kwargs):
+    """
+    Only transforms on axis=0
+    """
+    cdef:
+        int tiebreak
+        Py_ssize_t i, j, N, K
+        int64_t val_start=0, grp_start=0, dups=0, sum_ranks=0, grp_vals_seen=1
+        int64_t grp_na_count=0
+        ndarray[int64_t] _as
+        ndarray[object] _values
+        bint pct, ascending, keep_na
+
+    tiebreak = tiebreakers[kwargs['ties_method']]
+    ascending = kwargs['ascending']
+    pct = kwargs['pct']
+    keep_na = kwargs['na_option'] == 'keep'
+    N, K = (<object> values).shape
+
+    masked_vals = np.array(values[:, 0], copy=True)
+    mask = missing.isnaobj(masked_vals)
+
+    if ascending ^ (kwargs['na_option'] == 'top'):
+        nan_fill_val = np.inf
+        order = (masked_vals, mask, labels)
+    else:
+        nan_fill_val = -np.inf
+        order = (masked_vals, ~mask, labels)
+    np.putmask(masked_vals, mask, nan_fill_val)
+    try:
+        _as = np.lexsort(order)
+    except TypeError:
+        # lexsort fails when missing data and objects are mixed
+        # fallback to argsort
+        _arr = np.asarray(list(zip(order[0], order[1], order[2])),
+                          dtype=[('values', 'O'), ('mask', '?'),
+                          ('labels', 'i8')])
+        _as = np.argsort(_arr, kind='mergesort', order=('labels',
+                                                        'mask', 'values'))
+
+    if not ascending:
+        _as = _as[::-1]
+
+    for i in range(N):
+        dups += 1
+        sum_ranks += i - grp_start + 1
+
+        if keep_na and (values[_as[i], 0] != values[_as[i], 0]):
+            grp_na_count += 1
+            out[_as[i], 0] = np.nan
+        else:
+            if tiebreak == TIEBREAK_AVERAGE:
```

Comment:
In general this looping mechanism isn't very efficient because it overwrites values each time within a loop over duplicated values. Given the benchmarks were still significantly faster I left as is and was planning to open up a separate change to optimize further, but could review as part of this if you feel this loop mechanism is not acceptable